Multiple Levels in Activities , Part 1

lunayach opened the pull request:
### Description
Added multiple levels in the following activities:

- Picture Puzzle Component
- Odd One Out Component
- Do's and Don'ts  Component
- Memory Game Component

Fixes #137

### Type of Change:


- Code
- User Interface
- Outreach
- This change requires a documentation update (software upgrade on readme file)
- New feature (non-breaking change which adds functionality pre-approved by mentors)



### How Has This Been Tested?
Wrote new tests in Karma and Mocha. Previous Karma and Mocha Tests passed.


### Checklist:


- [x] My PR follows the style guidelines of this project
- [x] I have performed a self-review of my own code or materials
- [x] I have commented my code or provided relevant documentation, particularly in hard-to-understand areas
- [x] Any dependent changes have been merged 
- [x] My changes generate no new warnings 
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been published in downstream modules

#### UI of the Navigate Level Component (middle Right) which would be used by all activities having multiple levels. (Only one activity is shown in GIF)
![final9](https://user-images.githubusercontent.com/21227893/42521138-aca7b0de-8485-11e8-837b-849fcf28690b.gif)

Buddhiprabha commented:
Can we increase the cell width of the memory game? The cells look cluttered as levels go up. Also the idea was to allow the user have the option to go to the next level or to the next next game. @Raxerz @shreyans29 @paavininanda

shreyans29 commented on config/passport.js:
```diff
@@ -76,7 +77,21 @@ module.exports = function(passport, models) {
+                    }, defaults: {
+                        user_id: user.id,
+                    }})
+                        .spread(() => {
```
Identation here..

shreyans29 replied:
. spread can be started from previous line itself..

lunayach replied:
Done.

shreyans29 commented on data/codes.json:
```diff
@@ -158,10 +158,30 @@
+            "reason": "Request body was empty"
+      }
     },
 
```
I think this line can be removed.

lunayach replied:
But like all the Error Codes have followed this pattern until now. So, this may not look consistent.

shreyans29 replied:
no no, I meant the empty line in middle

shreyans29 replied:
line 186

lunayach replied:
Done.

shreyans29 commented on src/app/ActivityConfig.ts:
```diff
@@ -1,47 +1,59 @@
+    levels: 3
   },
 ];
+
```
Nice design :)

shreyans29 commented on src/app/introduction/activity-2/activity-2.component.html:
```diff
@@ -9,5 +9,5 @@
     </div>
   </div>
-  <button-nav [completed]="completed" nexturl="/introduction/activity-3" prevurl="/introduction/activity-1"></button-nav>
+  <button-nav [completed]="completed" nexturl="/introduction/activity-3/1" prevurl="/introduction/activity-1"></button-nav>
```
No need to change previous url?

lunayach replied:
@shreyans29  Not for now, because multiple levels for those activities have not been done in this PR.

shreyans29 replied:
got it

Raxerz replied:
@lunayach please make note of such changes that need to be added in the next PR.
It is very easy to miss them.

shreyans29 commented on src/app/level-navigate/level-navigate.component.scss:
```diff
@@ -0,0 +1,18 @@
+  display: inline-block;
+  position: absolute;
+  top: 50%;
+  right: 0;
```
I think we have a variable for 0.. $zero I guess.

lunayach replied:
Currently, we did not have this variable. So added it.

shreyans29 commented on src/app/level-navigate/level-navigate.component.scss:
```diff
@@ -0,0 +1,18 @@
+  font-size: $font-size-small;
+}
+
+.active-level {
```
4 for css 2 for HTML

lunayach replied:
Okay, Got it. :+1:

shreyans29 commented on src/app/level-navigate/level-navigate.component.spec.ts:
```diff
@@ -0,0 +1,35 @@
+
+  beforeEach(async(() => {
+    TestBed.configureTestingModule({
+      declarations: [ LevelNavigateComponent ],
```
Space not needed.

lunayach replied:
Okay.

lunayach commented:
Addressed the comments.

shreyans29 commented on src/app/performance-display/performance-display.component.ts:
```diff
@@ -33,8 +34,11 @@ export class PerformanceDisplayComponent implements OnInit {
-    document.getElementById('message').focus();
+    this.currActivityURL = this.SuccessObject.currActivityURL;
+    this.nextActivityLevels = this.SuccessObject.nextActivityLevels;
+    for (let i = 2; i <= this.SuccessObject.levels; i++) {
```
add a semicolon there..

lunayach replied:
Done.

shreyans29 commented on src/app/meds-n-labels/activity-2/activity-2.component.ts:
```diff
@@ -36,26 +39,34 @@ export class MemoryGameComponent implements OnInit {
-                'rapid-diagnostic-test-1.png',
                 'virus-1.png',
             ];
+    public levelConfig = [{cardsCount: 8, rowCount: 4, columnCount: 2},
```
move things to new line after [

lunayach replied:
Done.

paavininanda commented on config/passport.js:
```diff
@@ -51,32 +52,46 @@ module.exports = function(passport, models) {
+                            return done(err);
+                        });
+                }
+
```
Extra line here also, I think this can be removed.

lunayach replied:
Done.

paavininanda commented on src/app/malaria-101/activity-3/activity-3.component.ts:
```diff
@@ -9,6 +9,9 @@ import { BadgeService } from '../../services/BadgeService/badge.service';
 import { PerformanceDisplayService } from '../../services/performance-display.service';
 
+import { ActivatedRoute } from '@angular/router';
+
```
Extra space here also

lunayach replied:
Done.

paavininanda commented on src/app/malaria-101/activity-3/activity-3.component.ts:
```diff
@@ -91,11 +106,10 @@ export class OddOneOutComponent implements OnInit {
             this._sharedData.customErrorAlert(this.alerts.activityFailMsg, this.alerts.activityFailTitle);
         }
-        this._obs = observableInterval(1000).pipe(
+        this._obs = observableInterval(10).pipe(
```
Why did you change this?

lunayach replied:
Here, 1000 was the time delay between displaying the next question, when the user has answered the previous one. So, this being not a smooth experience, I reduced the delay to 10ms.

paavininanda commented on src/app/malaria-101/activity-3/activity-3.component.ts:
```diff
@@ -32,9 +37,14 @@ export class OddOneOutComponent implements OnInit {
     public alerts: any;
     public solutions = '';
+    public level: number;
+    public levelConfig = [4, 6, 8]; // no of questions depending upon the level
```
Number of questions is okay, but don't we have separate questions for each level ?

lunayach replied:
Yup. Added.

shreyans29 commented on test/test.js:
```diff
@@ -291,6 +291,37 @@ describe('Testing APIs', function() {
+                    });
+            });
+
+
```
Do these tests cover all possible scenarios.

lunayach replied:
Done.

Raxerz commented on database/models/levelProgress.js:
```diff
@@ -0,0 +1,48 @@
+            unique: true
+        }
+    }, {
+        tableName: 'levelProgress'
```
table name change with underscore separation

lunayach replied:
Done.

Raxerz commented on src/app/introduction/activity-3/activity-3.component.ts:
```diff
@@ -364,11 +374,13 @@ export class PicturePuzzleComponent implements OnInit {
+          this._dashboardService.updateActivityLevel({activity: 'oddOneOut', level: this.level}).subscribe(() => {});
+          if (this.level === 1) {
+            const badgeNumber = 1;
+            const currStage = 2;
```
constant naming convention

lunayach replied:
Done in the previous PR.

Raxerz commented on src/app/level-navigate/level-navigate.component.html:
```diff
@@ -0,0 +1,9 @@
+<div id="level-navigate" *ngIf="levelsURL.length > 1">
+  <div *ngFor="let levelURL of levelsURL; index as i;">
+    <a [href]="levelURL" [ngClass]="{'active-level': (i+1) === userCurrentLevel }">
```
can the comparison operation be moved to the ts file?

lunayach replied:
To bind a different class to specific HTML element, we would need to do a comparison inline, this way. There may not be a .ts way for this.

Raxerz commented on src/app/malaria-101/activity-2/activity-2.component.ts:
```diff
@@ -16,6 +17,8 @@ export class DragdropComponent implements OnInit {
     public dropCheckDont: String;
     language: any;
+    private score = 0;
+    private pointsPerCorrectAnswer = 10;
```
constant naming convention

lunayach replied:
Done in the previous PR.

shreyans29 commented on src/app/introduction/activity-3/activity-3.component.ts:
```diff
@@ -19,7 +20,10 @@ export class PicturePuzzleComponent implements OnInit {
 
-    private PUZZLE_DIFFICULTY = 4;
+    private PUZZLE_DIFFICULTY: number;
+    private difficultyList = [2, 4, 6];
```
will this be changed ? if not make it constant. Same with levels.

paavininanda replied:
And also include this level list in some common file. Do the same for all activities

lunayach replied:
Done. for making **readonly**

For storing in some common file, may not work here. Because, the way levels are increased in different activities is different. That is sometimes, it is just changing no. of fragments like in this case and then somtimes, it involves fetching a different property from a JSON file like objects marked under 'level1', 'level2' etc. Ex. for different set of questions for different levels.
So having a **readonly** activity-specific declaration at the top, may not be a bad idea.

shreyans29 commented on src/app/introduction/activity-3/activity-3.component.ts:
```diff
@@ -19,7 +20,10 @@ export class PicturePuzzleComponent implements OnInit {
     @ViewChild('video') video: any;
 
-    private PUZZLE_DIFFICULTY = 4;
+    private PUZZLE_DIFFICULTY: number;
```
This is not a read only variable. Please use camelCase.

lunayach replied:
Done.

paavininanda commented on src/app/malaria-101/activity-2/activity-2.component.ts:
```diff
@@ -25,6 +28,10 @@ export class DragdropComponent implements OnInit {
     public position= 'col-md-10 col-md-offset-2 introbody';
     public completed = false;
     public alerts: any;
+    public levelConfig = [5, 7, 10];
```
This can be stored in some common file.

lunayach commented:
@paavininanda  Actually, I have done changes locally. Would push them now.

lunayach commented:
Addressed comments.

Raxerz commented on src/app/introduction/activity-3/activity-3.component.ts:
```diff
@@ -19,7 +20,10 @@ export class PicturePuzzleComponent implements OnInit {
-    private PUZZLE_DIFFICULTY = 4;
+    private puzzleDifficulty: number;
+    private readonly DIFFICULTY_LIST = [4, 6, 8];
+    private levels = ['level1', 'level2', 'level3'];
```
can be a constant?

shreyans29 replied:
yep that and naming convention

lunayach replied:
Yes. This would change when the previous PR is merged.

Raxerz commented on src/app/introduction/activity-3/activity-3.component.ts:
```diff
@@ -57,9 +61,11 @@ export class PicturePuzzleComponent implements OnInit {
     public completed = false;
     public alerts: any;
     public userData: any;
+    public currStage = 2;
```
constant?

lunayach replied:
Done in the previous PR.

Raxerz commented on src/app/malaria-101/activity-3/activity-3.component.ts:
```diff
@@ -56,9 +66,15 @@ export class OddOneOutComponent implements OnInit {
             this._data = JSON.parse(response.data);
-            this.shuffle(this._data.quizlist);
+            this._data = this._data[this.levels[this.level - 1]];
+            console.log(JSON.stringify(this._data));
```
remove the log statement

lunayach replied:
Removed.

lunayach commented:
Addressed comments.

Buddhiprabha commented:
@lunayach why is @shreyans29 's review dismissed?

lunayach commented:
@Buddhiprabha even if we make one line change as compared to the approved version. Review gets dismissed. I didn't dismiss it. GitHub automatically does it.

paavininanda commented:
@lunayach Could you resolve the conflicts ?

shreyans29 commented on routes/api.js:
```diff
@@ -158,6 +159,66 @@ router.patch('/user/badge/update', authenticationHelpers.isAuthOrRedirect, (req,
+    }}, {raw: true})
+        .then(data => {
+            if (!data) {
+                return res.status(200).json({info: errorCode.PCE022.message, code: errorCode.PCE022.code});
```
error code..

lunayach replied:
@shreyans29 error code seems correct to me?

paavininanda replied:
He means to say that res.status should not be 200 since its an error. It should be 402.

lunayach replied:
Okay! Got it.

shreyans29 commented on routes/api.js:
```diff
@@ -158,6 +159,66 @@ router.patch('/user/badge/update', authenticationHelpers.isAuthOrRedirect, (req,
+            }}, {raw: true})
+                .then(activityData => {
+                    if (!activityData) {
+                        return res.status(200).json({info: errorCode.PCE036.message, code: errorCode.PCE036.code});
```
error code

lunayach replied:
For this one as well.

Raxerz commented on routes/api.js:
```diff
@@ -159,6 +160,66 @@ router.patch('/user/badge/update', authenticationHelpers.isAuthOrRedirect, (req,
+    }}, {raw: true})
+        .then(data => {
+            if (!data) {
+                return res.status(402).json({info: errorCode.PCE022.message, code: errorCode.PCE022.code});
```
402 error code?

shreyans29 replied:
I think it should be 400

Raxerz commented on routes/api.js:
```diff
@@ -159,6 +160,66 @@ router.patch('/user/badge/update', authenticationHelpers.isAuthOrRedirect, (req,
+            }}, {raw: true})
+                .then(activityData => {
+                    if (!activityData) {
+                        return res.status(402).json({info: errorCode.PCE036.message, code: errorCode.PCE036.code});
```
402 is again used? I don't think its right

shreyans29 replied:
yes it should be 400 as in the whole app.

Raxerz replied:
@shreyans29 I think it should be 200 only, the same is being discussed in the next PR.

lunayach replied:
@shreyans29, @Raxerz may be correct here. Because **200** actually means that **if** block worked properly. So, for a Developer, this would mean that exceptions are handled correctly and hence success.

Raxerz commented on src/app/introduction/activity-3/activity-3.component.ts:
```diff
@@ -369,10 +377,15 @@ export class PicturePuzzleComponent implements OnInit {
-          this._leaderBoardService.updateLeaderBoard({activity: this.ACTIVITY, level: 'level1'})
+          this._leaderBoardService.updateLeaderBoard({activity: this.ACTIVITY, level: this.LEVELS[this.level - 1]});
+
+          this._dashboardService.updateActivityLevel({activity: 'oddOneOut', level: this.level}).subscribe(() => {});
```
can 'oddOneOut' be made a constant?

lunayach replied:
Done.

Raxerz replied:
I don't see this change made

shreyans29 commented on app.js:
```diff
@@ -29,7 +29,7 @@ winston.add(
 
 // Sync Database
-models.sequelize.sync().then(function() {
+models.sequelize.sync({force: true}).then(function() {
```
if this is what you referred to that day, please remove it.  It should not be the the branch. You can keep using it for dev purposes.

lunayach replied:
Okay! Corrected it.

lunayach commented on src/app/meds-n-labels/activity-2/activity-2.component.ts:
```diff
@@ -166,15 +180,19 @@ export class MemoryGameComponent implements OnInit {
-                  this._performanceService.openDialog(this.CURR_STAGE); }
-                  this._leaderBoardService.updateLeaderBoard({activity: this.ACTIVITY, level: 'level1'});
+                  this._leaderBoardService.updateLeaderBoard({activity: this.ACTIVITY, level: this.LEVELS[this.levelNumber - 1]});
+                  this._dashboardService.updateActivityLevel({activity: 'memoryGame', level: this.levelNumber}).subscribe(() => {});
```
Done.

lunayach commented on src/app/malaria-101/activity-3/activity-3.component.ts:
```diff
@@ -147,8 +157,11 @@ export class OddOneOutComponent implements OnInit {
             }).then( res => { if (!this.completed) {
-              this._badgeService.updateBadgeNumber(this.BADGE_NUMBER).subscribe(response => response);
-              this._performanceService.openDialog(this.CURR_STAGE);
+              this._dashboardService.updateActivityLevel({activity: 'oddOneOut', level: this.level}).subscribe(() => {});
```
Done.

Raxerz commented:
@lunayach the activities that have been changed are not working for some reason.
![screenshot from 2018-08-01 23 07 49](https://user-images.githubusercontent.com/16686007/43538475-c68d5664-95df-11e8-9191-d78a0e9710af.png)
![screenshot from 2018-08-01 23 06 53](https://user-images.githubusercontent.com/16686007/43538487-cf6a465c-95df-11e8-857c-ac58b6c05be1.png)

lunayach commented:
@Raxerz please check now. Actually, there was just a mistake in naming variables. :sweat_smile:

Raxerz commented:
@lunayach It works fine now but there seems to be issues. Please check the following as well.
I have selected the wrong options but I still get a success notification
![screenshot from 2018-08-03 01 10 25](https://user-images.githubusercontent.com/16686007/43607120-ef6f70cc-96ba-11e8-8388-b3c991b62369.png)

The options did not appear at certain times, this was inconsistent but usually happened when I answered the questions too quickly.
![screenshot from 2018-08-03 01 10 59](https://user-images.githubusercontent.com/16686007/43607144-03a45e0e-96bb-11e8-9a14-737fb346e50f.png)

I got the success message but the level navigator arrow to move to the next level didn't appear
![screenshot from 2018-08-03 01 11 27](https://user-images.githubusercontent.com/16686007/43607269-5c731a3e-96bb-11e8-9592-0778ee430f05.png)

Didn't we decide on having level 3 as level 1 here?
![screenshot from 2018-08-03 01 13 52](https://user-images.githubusercontent.com/16686007/43607308-7a3a1e6e-96bb-11e8-82ec-cc1a16a4b9cb.png)

Raxerz commented:
@lunayach It works fine now but there seems to be issues. Please check the following as well.
I have selected the wrong options but I still get a success notification
![screenshot from 2018-08-03 01 10 25](https://user-images.githubusercontent.com/16686007/43607493-fcc8628c-96bb-11e8-93c0-b13da3c83b50.png)

The options did not appear at certain times, this was inconsistent but usually happened when I answered the questions too quickly.
![screenshot from 2018-08-03 01 10 59](https://user-images.githubusercontent.com/16686007/43607513-0b52e00c-96bc-11e8-9a96-a7cf6beb37f3.png)

I got the success message but the level navigator arrow to move to the next level didn't appear
![screenshot from 2018-08-03 01 11 27](https://user-images.githubusercontent.com/16686007/43607532-14da527c-96bc-11e8-8a61-adcc2d8d8d14.png)

The image is cutting off the border here and make the changes related to the size as suggested by buddhi.
![screenshot from 2018-08-03 01 12 53](https://user-images.githubusercontent.com/16686007/43607598-3db131ca-96bc-11e8-8363-00fabf4eb3a3.png)

Didn't we decide on having level 3 as level 1 here, I don't see that change made, have you made it in the second PR?
![screenshot from 2018-08-03 01 13 52](https://user-images.githubusercontent.com/16686007/43607605-46a790e4-96bc-11e8-8017-8291d5bc234d.png)

shreyans29 commented:
@lunayach please address the issues @Raxerz  has mentioned asap.. there are lots of PRs still pending..

lunayach commented:
@Raxerz done the changes.
For **1**, It comes when you use all available sentences. We allow him to go to the next activity even if his all options are not correct. We do keep a track of no. of correct answers and accordingly, he gets a dynamic score in LeaderBoard.

For **2**, There was just a problem in renaming a variable. Fixed it.
For **3**, It's a desirable behaviour. Please click **next** to complete next set and completing that would complete the activity. Anyway, this behaviour is changed in the next PR.
For **4, 5** Fixed border and added those images.

Raxerz commented:
@lunayach for **1** how will the user know if he was right or not? This is very misleading.

Raxerz commented on src/app/meds-n-labels/activity-2/activity-2.component.spec.ts:
```diff
@@ -28,6 +30,7 @@ describe('MemoryGameComponent', () => {
   let apiService: APIService;
   let sharedService: SharedDataService;
   let languageService: LanguageService;
+  const mockLevel = 1;
```
constant naming convention

Raxerz commented on src/app/meds-n-labels/activity-2/activity-2.component.ts:
```diff
@@ -77,6 +95,14 @@ export class MemoryGameComponent implements OnInit {
+        this._route.params.subscribe( params => {
+        this.currentLevel = this.LEVEL_CONFIG[ params.level - 1];
+        this.levelNumber =  parseInt(params.level, 10);
+          if (this.levelNumber === 3) {
```
make the 3 a constant

Raxerz commented:
@paavininanda please check and see if this PR is fine so that it can be merged.
